start.py

```diff
@@ -805,6 +801,40 @@ def start_app(m2ee):
         logger.warning('start failed, stopping')
         sys.exit(1)
 
+@atexit.register
+def terminate_process():
```

Comment:
I don't think we should add the extra dependency (mostly because of shipping the .so file which will break if we switch python/linux versions).

Instead, we can kill the entire process group of the current process. I believe process groups were invented for purposes such as this. Example code:
```
>>> import os
>>> import signal
>>> os.killpg(os.getpgid(0), signal.SIGTERM)
```

Example process ids and groups for a container that we have:
```
vcap@8f18f639-cce0-4a2e-4745-18d18fa28d76:~$ ps  xao pid,ppid,pgid,sid,comm
  PID  PPID  PGID   SID COMMAND
    1     0     1     1 runc:[2:INIT]
    7     0     7     7 python
   11     0    11    11 diego-sshd
   60     7     7     7 sed
   62     0    61    61 java
  146     7     7     7 nginx
  147   146     7     7 nginx
 6845    11  6845  6845 bash
 6870  6845  6870  6845 ps
```

Reply:
> `os.killpg(os.getpgid(0), signal.SIGTERM)`

This is a very strong and risky assumption that we always run a *containerized* environment such that we can kill parent process ID `0`.

Reply:
@jtwaleson All the comments were applied.

However, the test case for this PR consistently now takes the longest although we use a configured `start_timeout=15` with CF. Something to note down for later maybe.

Reply:
@jtwaleson @xiwenc 

Although the above changes are applied, there's something in the current code that needs to be observed. 

- Method `terminate_process` is registered as an exit handler for `sys`.
- Method `sigterm_handler` is registered as a SIGTERM handler [here](https://github.com/mendix/cf-mendix-buildpack/blob/master/start.py#L966).

Situation with `psutil`:

- :+1: A recursive list of child processes are obtained and ensured that all child processes are terminated and then `start.py` exists with `1`.
- :-1: A dependency is pulled in specifically for Linux that will not be portable if this buildpack is used in other platforms.

Situation with `os`:

- :+1: No external dependency.
- :+1: Send a standard `SIGTERM` to process group of `start.py`.
- :-1: Redundantly catch the same signal in the signal handler and do an exit `0`. This causes to lose the terminal exit code which is ` != 0`.

Using `os` creates a loop of logic because of using process group although it's simple. The API does not provide a way to exclude the parent process of the group. If we want to be precisely correct, the road map would be to implement a method which becomes similar to that of `psutil`.